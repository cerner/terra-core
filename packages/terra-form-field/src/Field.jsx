import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import classNames from 'classnames/bind';
import IconError from 'terra-icon/lib/icon/IconError';
import styles from './Field.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * The form control elements the Field contains.
   */
  children: PropTypes.node,
  /**
   * Error message for when the input is invalid. This will only be displayed if isInvalid is true.
   */
  error: PropTypes.node,
  /**
   * Error Icon to display when the input is invalid.
   */
  errorIcon: PropTypes.element,
  /**
   * Help element to display with the input.
   */
  help: PropTypes.node,
  /**
   * Whether or not to hide the required indicator on the label.
   */
  hideRequired: PropTypes.bool,
  /**
   * The htmlFor attribute on the field label.
   */
  htmlFor: PropTypes.string,
  /**
   * Whether the field displays as Invalid. Use when value does not meet validation pattern.
   */
  isInvalid: PropTypes.bool,
  /**
   * Whether or not the field is an inline field.
   */
  isInline: PropTypes.bool,
  /**
   * Whether or not the label is visible. Use this props to hide a label while still creating it on the DOM for accessibility.
   */
  isLabelHidden: PropTypes.bool,
  /**
   * The label of the form control children.
   */
  label: PropTypes.node.isRequired,
  /**
   * Attributes to attach to the label.
   */
  // eslint-disable-next-line react/forbid-prop-types
  labelAttrs: PropTypes.object,
  /**
   * Set the max-width of a field using `length` or `%`.  Best practice recommendation to never exceed
   * a rendered value of 1020px. _(Note: Providing custom inline styles will take precedence.)_
   */
  maxWidth: PropTypes.string,
  /**
   * Whether or not the field is required.
   */
  required: PropTypes.bool,
  /**
   * Whether or not to append the 'optional' label to a non-required field label.
   */
  showOptional: PropTypes.bool,
  /**
   *  Provides first class prop for custom inline styles
   */
  // eslint-disable-next-line react/forbid-prop-types
  style: PropTypes.object,
};

const defaultProps = {
  children: null,
  error: null,
  errorIcon: <IconError />,
  help: null,
  hideRequired: false,
  htmlFor: undefined,
  isInvalid: false,
  isInline: false,
  isLabelHidden: false,
  labelAttrs: {},
  maxWidth: undefined,
  required: false,
  showOptional: false,
};

const hasWhiteSpace = s => /\s/g.test(s);
// Detect IE 10 or IE 11
// TODO - Delete detection logic when we drop support for IE
const isIE = () => (window.navigator.userAgent.indexOf('Trident/6.0') > -1 || window.navigator.userAgent.indexOf('Trident/7.0') > -1);

const Field = (props) => {
  const {
    children,
    error,
    errorIcon,
    help,
    hideRequired,
    htmlFor,
    isInvalid,
    isInline,
    isLabelHidden,
    label,
    labelAttrs,
    maxWidth,
    required,
    showOptional,
    style,
    ...customProps
  } = props;

  const customStyles = maxWidth ? ({ maxWidth, ...style }) : style;

  const fieldClasses = cx([
    'field',
    { 'field-inline': isInline },
    customProps.className,
  ]);

  const labelClassNames = cx([
    'label',
    labelAttrs.className,
  ]);

  // Checks to run when not in production
  // TODO - Remove this once we make this a required prop.
  // More info: https://github.com/cerner/terra-core/issues/2290
  if (process.env.NODE_ENV !== 'production') {
    if (!htmlFor) {
      // eslint-disable-next-line
      console.warn('The htmlFor prop will be required in the next major version bump of terra-form-field. It is needed for creating an accessible mapping from the form field to its related error and help text.');
    }

    if (htmlFor && hasWhiteSpace(htmlFor)) {
      // eslint-disable-next-line
      console.warn('The htmlFor prop should be a string without white spaces as it will be used as an HTML attribute value. Use - or _ in place of white space characters.');
    }
  }

  /**
   * IE + JAWS has trouble reading aria-describedby content with our form components.
   * Using feature detect for Microsoft browsers and injecting the help and error messages
   * into the label as visually hidden text so JAWS can announce them correctly in IE.
   */
  const IEDescriptionText = (
    isIE()
      ? (
        <div className={cx('visually-hidden-text')}>
          {isInvalid && error ? error : null}
          {help}
        </div>
      ) : null
  );

  const content = React.Children.map(children, (child) => {
    if ((required || isInvalid) && child && (child.type.isInput || child.type.isSelect || child.type.isTextarea)) {
      return React.cloneElement(child, {
        ...required && { required: true },
        ...isInvalid && { isInvalid: true },
      });
    }
    return child;
  });

  const labelGroup = (
    <div className={cx(['label-group', { 'label-group-hidden': isLabelHidden }])}>
      {isInvalid && <div className={cx('error-icon')}>{errorIcon}</div>}
      {
        <label htmlFor={htmlFor} {...labelAttrs} className={labelClassNames}>
          {required && (isInvalid || !hideRequired) && <div className={cx('required')}>*</div>}
          {label}
          {required && !isInvalid && hideRequired && <div className={cx('required-hidden')}>*</div>}
          {showOptional && !required
            && (
              <FormattedMessage id="Terra.form.field.optional">
                {optionalText => (
                  <div className={cx('optional')}>{optionalText}</div>
                )}
              </FormattedMessage>
            )}
          {IEDescriptionText}
        </label>
      }
      {!isInvalid && <div className={cx('error-icon-hidden')}>{errorIcon}</div>}
    </div>
  );

  /* eslint-disable react/forbid-dom-props */
  return (
    <div style={customStyles} {...customProps} className={fieldClasses}>
      {labelGroup}
      {content}
      {isInvalid && error && <div aria-live="assertive" tabIndex="-1" id={htmlFor ? `${htmlFor}-error` : undefined} className={cx('error-text')}>{error}</div>}
      {help && <div tabIndex="-1" id={htmlFor ? `${htmlFor}-help` : undefined} className={cx('help-text')}>{help}</div>}
    </div>
  );
  /* eslint-enable react/forbid-dom-props */
};

Field.propTypes = propTypes;
Field.defaultProps = defaultProps;

export default Field;
