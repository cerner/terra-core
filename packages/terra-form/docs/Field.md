# Field

Generic form field component which handles the layout of a standard form field including errors, help text, label, value and widget placement.

## Usage

```jsx
import React from 'react';
import { Field } from 'terra-form';

<Field
  type="checkbox"
  label="Do you have any Children?"
  name="children_present"
  value="children_present"
  error="This field is required"
  help="Families are eligible for family package plans"
  required={true}
  attrs={{ className: 'healtheintent-application' }}
  isInline={false}
 >
   {this.props.children}
</Field>
```
