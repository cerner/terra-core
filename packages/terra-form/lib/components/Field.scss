@import '~terra-mixins';

.terra-Form-field {
  margin-bottom: 1.428em;
}

@media screen and (min-width: $terra-small-breakpoint) {
  .terra-Form-field--inline {
    display: inline-block;
    vertical-align: top;
  }
}

.terra-Form-field--inline {
  @include terra-margin-end(0.7143em);
}

.terra-Form-label {
  display: block;
  font-weight: 700;
  margin-bottom: 0.357em;
}

.terra-Form-helpText {
  display: block;
  margin-top: 0.357em;
}

.terra-Form-error {
  color: $terra-red-70;
  display: block;
  margin-top: 0.357em;
}

.terra-Form-field--required > label:first-child {
  &::before {
    @include terra-padding-end(0.357em);
    color: $terra-red-70;
    content: '*';
    display: inline-block;
  }
}
