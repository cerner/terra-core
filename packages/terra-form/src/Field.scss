@import '~terra-mixins';

.terra-Form-field,
.terra-Form-fieldset {
  border: 0;
  margin-bottom: 1.428em;
  padding: 0;
}

@media screen and (min-width: 544px) {
  .terra-Form-field--inline,
  .terra-Form-fieldset--inline {
    @include terra-margin-end(0.7143em);
    display: inline-block;
    vertical-align: top;
  }
}

.terra-Form-field--required > .terra-Form-label:first-child,
.terra-Form-fieldset--required > .terra-Form-legend:first-child {
  &::before {
    @include terra-padding-end(0.357em);
    color: #e50000;
    content: '*';
    display: inline-block;
  }
}

.terra-Form-label {
  display: block;
  font-weight: bold;
  margin-bottom: 0.357em;
}

.terra-Form-legend {
  border: 0;
  display: block;
  font-weight: bold;
  margin-bottom: 0.357em;
  max-width: 100%; // IE11 fix to ensure legends wrap
  padding: 0;
  width: auto;
}

.terra-Form-helpText {
  display: block;
  margin-bottom: 0.357em;
  outline: 0;
}

.terra-Form-error {
  color: #e50000;
  display: block;
  margin-bottom: 0.357em;
  outline: 0;
}

.terra-Form-fieldset-children {
  margin-top: 0.714em;
}
