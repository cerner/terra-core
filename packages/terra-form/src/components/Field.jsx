/* eslint-disable react/jsx-boolean-value, jsx-a11y/label-has-for */
import React, { PropTypes } from 'react';
import classNames from 'classnames';

const propTypes = {
  /**
   * Children the Field contains
   */
  children: PropTypes.node,
  /**
   * Error message for when the input is invalid
   */
  error: PropTypes.node,
  /**
   * Help element to display with the input
   */
  help: PropTypes.node,
  /**
   * Determines whether the field in an inline field
   */
  isInline: PropTypes.bool,
  /**
   * Label of the input
   */
  label: PropTypes.node,
};

const defaultProps = {
  children: null,
  error: null,
  help: null,
  isInline: false,
  label: null,
};

const Field = ({ children, error, help, isInline, label, ...customProps }) => {
  const fieldClasses = classNames(
    'terra-Form-field',
    { 'terra-Form-field--inline': isInline },
    customProps.className,
  );

  return (
    <p {...customProps} className={fieldClasses}>
      {label && <label className="terra-Form-label">{label}</label>}
      {children}
      {help && <small className="terra-Form-helpText">{help}</small>}
      {error && <small className="terra-Form-error" tabIndex="-1">{error}</small>}
    </p>
  );
};

Field.propTypes = propTypes;
Field.defaultProps = defaultProps;

export default Field;
