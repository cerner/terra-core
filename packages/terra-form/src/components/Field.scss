@import '~terra-mixins';
@import '../variables';

.terra-Form-field,
.terra-Form-fieldset {
  border: $terra-form-field-border;
  margin-bottom: 1.428em;
}

@media screen and (min-width: $terra-small-breakpoint) {
  .terra-Form-field--inline,
  .terra-Form-fieldset--inline {
    display: inline-block;
    vertical-align: top;
  }
}

.terra-Form-field--required > .terra-Form-label:first-child,
.terra-Form-fieldset--required > .terra-Form-legend:first-child {
  &::before {
    @include terra-padding-end(0.357em);
    color: $terra-form-required-color;
    content: $terra-form-required-indicator;
    display: inline-block;
  }
}

.terra-Form-field--inline,
.terra-Form-fieldset--inline {
  @include terra-margin-end(0.7143em);
}

.terra-Form-label {
  display: block;
  font-weight: $terra-form-label-font-weight;
  margin-bottom: 0.357em;
}

.terra-Form-legend {
  border: 0;
  color: $terra-form-legend-color;
  display: block;
  font-weight: bold;
  margin-bottom: 0.357em;
  max-width: 100%; // IE11 fix to ensure legends wrap
  padding: 0;
  width: auto;
}

.terra-Form-helpText {
  display: block;
  margin-top: 0.357em;
  outline: 0;
}

.terra-Form-error {
  color: $terra-form-error-color;
  display: block;
  margin-top: 0.357em;
  outline: 0;
}

