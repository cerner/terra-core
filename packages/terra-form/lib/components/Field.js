'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

require('terra-base/lib/baseStyles');

require('./Field.scss');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; } /* eslint-disable react/jsx-boolean-value, jsx-a11y/label-has-for */


var propTypes = {
  /**
   * Children the Field contains
   */
  children: _react.PropTypes.node,
  /**
   * Error message for when the input is invalid
   */
  error: _react.PropTypes.node,
  /**
   * Help element to display with the input
   */
  help: _react.PropTypes.node,
  /**
   * Determines whether the field in an inline field
   */
  isInline: _react.PropTypes.bool,
  /**
   * Label of the input
   */
  label: _react.PropTypes.node,
  /**
   * Determines whether the field is required
   */
  required: _react.PropTypes.bool
};

var defaultProps = {
  children: null,
  error: null,
  help: null,
  isInline: false,
  label: null,
  required: false
};

var Field = function Field(_ref) {
  var children = _ref.children,
      error = _ref.error,
      help = _ref.help,
      isInline = _ref.isInline,
      label = _ref.label,
      required = _ref.required,
      customProps = _objectWithoutProperties(_ref, ['children', 'error', 'help', 'isInline', 'label', 'required']);

  var fieldClasses = (0, _classnames2.default)('terra-Form-field', { 'terra-Form-field--inline': isInline }, { 'terra-Form-field--required': required }, customProps.className);

  return _react2.default.createElement(
    'p',
    _extends({}, customProps, { className: fieldClasses }),
    label && _react2.default.createElement(
      'label',
      { className: 'terra-Form-label' },
      label
    ),
    children,
    help && _react2.default.createElement(
      'small',
      { className: 'terra-Form-helpText' },
      help
    ),
    error && _react2.default.createElement(
      'small',
      { className: 'terra-Form-error', tabIndex: '-1' },
      error
    )
  );
};

Field.propTypes = propTypes;
Field.defaultProps = defaultProps;

exports.default = Field;