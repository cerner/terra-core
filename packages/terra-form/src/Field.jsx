import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import 'terra-base/lib/baseStyles';

import './Field.scss';

const propTypes = {
  /**
   * Children the Field contains
   */
  children: PropTypes.node,
  /**
   * Error message for when the input is invalid
   */
  error: PropTypes.node,
  /**
   * Help element to display with the input
   */
  help: PropTypes.node,
  /**
   * The htmlFor attribute on the field label.
   */
  htmlFor: PropTypes.string,
  /**
   * Determines whether the field is an inline field
   */
  isInline: PropTypes.bool,
  /**
   * Label of the input
   */
  label: PropTypes.node,
  /**
   * Determines whether the field is required
   */
  required: PropTypes.bool,
};

const defaultProps = {
  children: null,
  error: null,
  help: null,
  htmlFor: undefined,
  isInline: false,
  label: null,
  required: false,
};

const Field = ({ children, error, help, htmlFor, isInline, label, required, ...customProps }) => {
  const fieldClasses = classNames(
    'terra-Form-field',
    { 'terra-Form-field--inline': isInline },
    { 'terra-Form-field--required': required },
    customProps.className,
  );

  return (
    <div {...customProps} className={fieldClasses}>
      {label && <label className="terra-Form-label" htmlFor={htmlFor}>{label}</label>}
      {children}
      {help && <small className="terra-Form-helpText" tabIndex="-1">{help}</small>}
      {error && <small className="terra-Form-error" tabIndex="-1">{error}</small>}
    </div>
  );
};

Field.propTypes = propTypes;
Field.defaultProps = defaultProps;

export default Field;
