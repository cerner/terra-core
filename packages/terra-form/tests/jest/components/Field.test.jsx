/* eslint-disable react/jsx-boolean-value */

import React from 'react';
import Field from '../../../src/components/Field';

it('should render a default component', () => {
  const field = (<Field />);
  const wrapper = shallow(field);
  expect(wrapper).toMatchSnapshot();
});

it('should render a Field that has the error, label, and help props set', () => {
  const input = (
    <Field
      required={false}
      name="foo"
      value="bar"
    />);

  const wrapper = shallow(input);
  expect(wrapper).toMatchSnapshot();
});

it('should render a Field that has children in it', () => {
  const input = (
    <Field
      required={false}
      name="foo"
      value="bar"
    >
      <input type="text" value="Test" />
    </Field>
  );

  const wrapper = shallow(input);
  expect(wrapper).toMatchSnapshot();
});
