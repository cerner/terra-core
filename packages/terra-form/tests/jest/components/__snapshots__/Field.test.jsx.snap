exports[`test should render a Field that has children in it 1`] = `
<p
  className="terra-Form-field"
  name="foo"
  value="bar">
  <input
    type="text"
    value="Test" />
</p>
`;

exports[`test should render a Field that has the error, label, and help props set 1`] = `
<p
  className="terra-Form-field"
  name="foo"
  value="bar" />
`;

exports[`test should render a default component 1`] = `
<p
  className="terra-Form-field" />
`;
